### ISSUE TITLE
Composer Support

### ISSUE THREAD

**Issue by robertmain** [WP-for-Church/Sermon-Manager#199]
We could look to using composer for managing our dependencies etc. That way we could get a lot of 3rd party code out of the repo. Even things like CMB etc. could be moved out of the repo and grabbed as composer dependencies until we factor it out.

At the very least we could use it to install PHPCS, PHPUnit and the WP PHPCS Ruleset etc.

**Comment by robertmain** [comment]
I've actually just realised that this may not be possible due to the way that WordPress works. In order for this to work, we would need to install the composer dependencies when the plugin was installed. In order to do that we'd need to have composer installed when the user installed the plugin, so I'm not sure if this is actually viable..

**Comment by nikola3244** [comment]
Hahah, I've already thought of it. Do a step back - dependencies could be installed during the package building phase. :wink: 

Our CI system has a scenario for package release, I would just need to add a line or two for composer init

**Comment by robertmain** [comment]
When you release a WordPress plugin, is the code for the plugin installed on GH or is it posted to an external package host?

**Comment by nikola3244** [comment]
It's on wp.org SVN. GitHub is not required at all

**Comment by robertmain** [comment]
Right, so you'd have to have your CI/CD system make one commit for the code, then another commit for the composer deps? Kinda messy, isn't it?

**Comment by nikola3244** [comment]
Commit? No. 

This is the process:

1. Things get developed on `dev`, and on each commit on `dev`, CI system creates a development zip, and uploads it to wordpress.org. It's usually provided for users to test bug fixes (which can be downladed from [here](https://downloads.wordpress.org/plugin/sermon-manager-for-wordpress.zip))
2. Once a release is ready, `release/x.x.x` is branched from `dev` (locally, usually), version gets bumped, final testing happens, last bugs are fixed, etc, etc. When that's ready, it gets merged into `master` and `dev`. CI then makes a release package from `master`, uploads it to wordpress.org and tags it

If we opt for Composer (and I have talked with a teammate, we both would like to implement it into our products), I would just need to add a line or two of code to the Bash script that's executed when there's a commit to `master`.

I'm not sure what you mean by `one commit for the code, then another commit for the composer deps`.

**Comment by robertmain** [comment]
> I'm not sure what you mean by one commit for the code, then another commit for the composer deps.



I was thinking you'd need one commit to GitHub to trigger the CI pipeline. This in turn would download the dependencies and trigger another commit to svn or whatever. Honestly, why wp is still stuck on SVN is beyond me...

Either way, the phrase "gl;hf" comes to mind 🙃

**Comment by nikola3244** [comment]
Either way, one commit on GitHub and one commit on SVN is needed. It doesn't matter if Composer is involved or not

**Comment by robertmain** [comment]
So you don't use any kind of GitHub to SVN bridge then?

**Comment by nikola3244** [comment]
Nope, just standard `git` and `svn`

**Comment by robertmain** [comment]
Do you mirror the commit messages over too?

**Comment by nikola3244** [comment]
Yup :slightly_smiling_face: 
https://plugins.trac.wordpress.org/log/sermon-manager-for-wordpress/
